README.md
# `tube-info`

Node.js command line app for exporting video/playlist data from a YouTube to a JSON/CSV file with extra features.

This app lets you download the [`metadata of each video`](#exportable-items) (e.g. video title, URL, uploader, and etc.) from a public/unlisted YouTube playlist and saves it to a JSON or CSV file.

![Demo of usage](./assets/images/demo-usage.gif)

## Features

- Highly customizable
- Promise based HTTP client using [`axios`](https://www.npmjs.com/package/axios)
- Terminal string styling using [`chalk`](https://www.npmjs.com/package/chalk)
- command-line interfaces & implements a help system using [`commander`](https://www.npmjs.com/package/commander)
- Simple config handling using [`conf`](https://www.npmjs.com/package/conf)
- JavaScript code linting is done using [`eslint`](https://www.npmjs.com/package/eslint)
- common interactive command line user interfaces using [`inquirer`](https://www.npmjs.com/package/inquirer)
- validate user input path using [`is-absolute`](https://www.npmjs.com/package/is-absolute)
- Converts json into csv using [`json2csv`](https://www.npmjs.com/package/json2csv)
- Elegant terminal spinner using [`ora`](https://www.npmjs.com/package/ora)
- Delightful JavaScript Testing using [`jest`](https://www.npmjs.com/package/jest)
- create build version using [`babel`](https://babeljs.io/)
- Has [`.editorconfig`](https://editorconfig.org/) which helps developers define and maintain consistent coding styles between different editors and IDEs.

> **Note** This app does **not** download videos from a YouTube playlist. It only downloads **text-based metadata** of videos from a playlist.

## Installation

> Before start [`node.js`](https://nodejs.org/en/) must pre-install in your machine.
> to verify installation, use `node -v` it show something like that `v16.15.0` see [`node.js`](https://nodejs.org/en/) for more details and how to install.

**note** for Linux and Mac users need to execute the following command with `sudo`

1. Install the app:

   ```shell
   npm install -g tube-info
   ```

2. Run the following command in Terminal to confirm that it's properly installed:

   ```shell
   tube-info --help
   ```

3. Get a YouTube API v3 key for free:

   - 3 min. YouTube tutorial - [`How to Get YouTube API Key 2021`](https://youtu.be/N18czV5tj5o)
   - RapidAPI blog article - [`How To Get a YouTube API Key (in 7 Simple Steps)`](https://rapidapi.com/blog/how-to-get-youtube-api-key/)

4. Run the following command and follow the on-screen instructions to register the YouTube API key:

   ```shell
   tube-info key
   ```

## Usage 🚀

Use the following command to get more info

```shell
tube-info --help
```

This will get the following results:

```txt
Usage: tube-info [options] [command]

Exports video/playlist data from YouTube to JSON/CSV file.


Options:
  -v, --version                   output the version number
  -h, --help                      display help for command

Commands:
  key                             Manage your YouTube API key.
  config [options]                Edit configurations of tube-info app.
  playlist [options] <ID Or URL>  Export playlist metadata its id or url.
  video [options] <ID Or URL>     Export video metadata by id or url.
  help [command]                  display help for command
```

### Export video

Export a YouTube video given its video ID, Or using video url.

```shell
tube-info video <videoId>
```

To find the video ID, go to the video homepage and copy the value of the `list` parameter in the URL `https://www.youtube.com/watch?v=[videoId]`).

For example, the [`Learn Typescript In Arabic 2022 - #01 - Introduction And What Is TypeScript`](https://www.youtube.com/watch?v=yUndnE-2osg) video has a video ID of `yUndnE-2osg`.

![Demo of video "id" command](./assets/images/demo-video-id.png)

#### Flags

- `-d`/`--default` - Skip all prompt questions and use the default configurations

#### Exportable Items

| Item           | JSON/CSV Key Name |
| -------------- | ----------------- |
| Channel id     | `channelId`       |
| Channel title  | `channelTitle`    |
| Comment count  | `commentCount`    |
| Description    | `description`     |
| Favorite count | `favoriteCount`   |
| Length         | `length`          |
| Likes          | `likeCount`       |
| Privacy status | `privacyStatus`   |
| Publish date   | `publishDate`     |
| Tags           | `tags`            |
| Thumbnail      | `thumbnail`       |
| Title          | `title`           |
| Url            | `url`             |
| View count     | `viewCount`       |

### Export Playlist

Export a YouTube playlist given its playlist ID, Or using playlist url.

```shell
tube-info playlist <playlistId>
```

To find the playlist ID, go to the playlist homepage and copy the value of the `list` parameter in the URL `https://www.youtube.com/playlist?list=[playlistId]`).

For example, the [`Learn Typescript 2022`](https://www.youtube.com/playlist?list=PLDoPjvoNmBAy532K9M_fjiAmrJ0gkCyLJ) playlist has a playlist ID of `PLDoPjvoNmBAy532K9M_fjiAmrJ0gkCyLJ`.

![Demo of "id" command](./assets/images/demo-playlist-id.png)

#### Flags

- `-d`/`--default` - Skip all prompt questions and use the default configurations

#### Exportable Items

| Item                      | JSON/CSV Key Name |
| ------------------------- | ----------------- |
| Position in the playlist  | `position`        |
| Title                     | `title`           |
| Uploader                  | `uploader`        |
| thumbnail                 | `thumbnail`       |
| Uploader URL              | `uploaderUrl`     |
| Description               | `description`     |
| Url                       | `url`             |
| Video privacy             | `videoPrivacy`    |
| Publish time (UTC)        | `publishTime`     |

**Note** Playlist position starts from 0

### Save Location

You can specify the absolute path of a folder where the files will be saved to. The default value is `~/tube-info`. Don't worry, the folder will be created if it doesn't exist.

Since it requires an **absolute** path, here are same examples:

- ✔️ `C:\Users\User\Downloads` (Windows)
- ✔️ `/usr/tube-info` (Mac/Linux)
- ❌  `../tube-info`
- ❌  `./tube-info`

### Example Output

#### Playlist Example

The app outputs the following when running `tube-info playlist PLDoPjvoNmBAy532K9M_fjiAmrJ0gkCyLJ`:

<details>
   <summary>JSON</summary>

```json
[
  {
    "position": 0,
    "thumbnail": "https://i.ytimg.com/vi/yUndnE-2osg/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #01 - Introduction And What Is TypeScript",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/yUndnE-2osg"
  },
  {
    "position": 1,
    "thumbnail": "https://i.ytimg.com/vi/pc5IlcEn8vw/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #02 - Install TypeScript And Transpile Files",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/pc5IlcEn8vw"
  },
  {
    "position": 2,
    "thumbnail": "https://i.ytimg.com/vi/CSll1rsRPOI/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #03 - Create Configuration And Watch Files",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/CSll1rsRPOI"
  },
  {
    "position": 3,
    "thumbnail": "https://i.ytimg.com/vi/OgxYA7G9HsM/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #04 - Statically vs Dynamically Typed Languages",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/OgxYA7G9HsM"
  },
  {
    "position": 4,
    "thumbnail": "https://i.ytimg.com/vi/quwf-YbyHVg/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #05 - Type Annotations And Any Data Type",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/quwf-YbyHVg"
  },
  {
    "position": 5,
    "thumbnail": "https://i.ytimg.com/vi/U405xMeS4lM/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #06 - Type Annotations With Arrays",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/U405xMeS4lM"
  },
  {
    "position": 6,
    "thumbnail": "https://i.ytimg.com/vi/n6JBmErg1OY/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #07 - Type Annotations With Multidimensional Arrays",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/n6JBmErg1OY"
  },
  {
    "position": 7,
    "thumbnail": "https://i.ytimg.com/vi/ibvt_Ala8wE/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #08 - Type Annotations With Function",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/ibvt_Ala8wE"
  },
  {
    "position": 8,
    "thumbnail": "https://i.ytimg.com/vi/IS2VuO0IWso/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #09 - Function Optional and Default Parameters",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/IS2VuO0IWso"
  },
  {
    "position": 9,
    "thumbnail": "https://i.ytimg.com/vi/RBOpzAQaQos/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #10 - Function Rest Parameter",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/RBOpzAQaQos"
  },
  {
    "position": 10,
    "thumbnail": "https://i.ytimg.com/vi/AWg__YvDdvg/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #11 - Type Annotations With Anonymous And Arrow Function",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/AWg__YvDdvg"
  },
  {
    "position": 11,
    "thumbnail": "https://i.ytimg.com/vi/TWTt63RJ3ic/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #12 - Data Types - Type Alias",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/TWTt63RJ3ic"
  },
  {
    "position": 12,
    "thumbnail": "https://i.ytimg.com/vi/J1WsNERYqYA/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #13 - Data Types - Type Alias Advanced",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/J1WsNERYqYA"
  },
  {
    "position": 13,
    "thumbnail": "https://i.ytimg.com/vi/BmmTQZsKXyw/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #14 - Data Types - Literal Types",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/BmmTQZsKXyw"
  },
  {
    "position": 14,
    "thumbnail": "https://i.ytimg.com/vi/hwTOswoq3BE/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #15 - Data Types - Tuple",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/hwTOswoq3BE"
  },
  {
    "position": 15,
    "thumbnail": "https://i.ytimg.com/vi/tPQCnP2IYn8/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #16 - Data Types - Void And Never",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/tPQCnP2IYn8"
  },
  {
    "position": 16,
    "thumbnail": "https://i.ytimg.com/vi/Lb8-2rI8nco/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #17 - Data Types - Enums Part 1",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/Lb8-2rI8nco"
  },
  {
    "position": 17,
    "thumbnail": "https://i.ytimg.com/vi/gMNoF-BVWpA/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #18 - Data Types - Enums Part 2",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/gMNoF-BVWpA"
  },
  {
    "position": 18,
    "thumbnail": "https://i.ytimg.com/vi/tPSHBw_2huc/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #19 - Data Types - Type Assertions",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/tPSHBw_2huc"
  },
  {
    "position": 19,
    "thumbnail": "https://i.ytimg.com/vi/Uf-ODplNJ7A/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #20 - Data Types - Union And Intersection Types",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/Uf-ODplNJ7A"
  },
  {
    "position": 20,
    "thumbnail": "https://i.ytimg.com/vi/7pjfbbZTOCU/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #21 - Type Annotations With Object",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/7pjfbbZTOCU"
  },
  {
    "position": 21,
    "thumbnail": "https://i.ytimg.com/vi/qopmfZ30_TQ/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #22 - Interface Declaration",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/qopmfZ30_TQ"
  },
  {
    "position": 22,
    "thumbnail": "https://i.ytimg.com/vi/ZOWwVhAUtZQ/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #23 - Interface Method And Parameters",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/ZOWwVhAUtZQ"
  },
  {
    "position": 23,
    "thumbnail": "https://i.ytimg.com/vi/7rJwXk_SuXQ/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #24 - Interface Reopen And Use Cases",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/7rJwXk_SuXQ"
  },
  {
    "position": 24,
    "thumbnail": "https://i.ytimg.com/vi/3iwthBqZwrM/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #25 - Interface Extend",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/3iwthBqZwrM"
  },
  {
    "position": 25,
    "thumbnail": "https://i.ytimg.com/vi/CaBIiFpxFTo/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #26 - Interface Final Discussion",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/CaBIiFpxFTo"
  },
  {
    "position": 26,
    "thumbnail": "https://i.ytimg.com/vi/mD3qHYKNBfo/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #27 - Class Type Annotations",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/mD3qHYKNBfo"
  },
  {
    "position": 27,
    "thumbnail": "https://i.ytimg.com/vi/FzROwTuVKr8/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #28 - Class Access Modifiers And Parameters Properties",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/FzROwTuVKr8"
  },
  {
    "position": 28,
    "thumbnail": "https://i.ytimg.com/vi/kXS5eprio7I/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #29 - Class Get And Set Accessors",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/kXS5eprio7I"
  },
  {
    "position": 29,
    "thumbnail": "https://i.ytimg.com/vi/mFnCSvduc8M/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #30 - Class Static Members",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/mFnCSvduc8M"
  },
  {
    "position": 30,
    "thumbnail": "https://i.ytimg.com/vi/klcmDKP_34s/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #31 - Class Implements Interface",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/klcmDKP_34s"
  },
  {
    "position": 31,
    "thumbnail": "https://i.ytimg.com/vi/nesuaOpNDWU/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #32 - Abstract Classes And Members",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/nesuaOpNDWU"
  },
  {
    "position": 32,
    "thumbnail": "https://i.ytimg.com/vi/V-CEhWbYDYY/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #33 - Polymorphism And Method Override",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/V-CEhWbYDYY"
  },
  {
    "position": 33,
    "thumbnail": "https://i.ytimg.com/vi/b1w1qa3H_vA/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #34 - Generics Introduction",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/b1w1qa3H_vA"
  },
  {
    "position": 34,
    "thumbnail": "https://i.ytimg.com/vi/BA40TsV6LxQ/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #35 - Generics Multiple Types",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/BA40TsV6LxQ"
  },
  {
    "position": 35,
    "thumbnail": "https://i.ytimg.com/vi/Mj6TJAoisxw/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #36 - Generics Classes",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/Mj6TJAoisxw"
  },
  {
    "position": 36,
    "thumbnail": "https://i.ytimg.com/vi/9uwul5pr0YE/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #37 - Generics And Interfaces",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/9uwul5pr0YE"
  },
  {
    "position": 37,
    "thumbnail": "https://i.ytimg.com/vi/Ot5AGaudAcg/sddefault.jpg",
    "title": "Learn Typescript In Arabic 2022 - #38 - The End And How To Master Typescript",
    "uploader": "Elzero Web School",
    "url": "https://youtu.be/Ot5AGaudAcg"
  }
]
```

</details>

<details>
   <summary>CSV</summary>

```csv
"position","title","uploader","url","thumbnail"
0,"Learn Typescript In Arabic 2022 - #01 - Introduction And What Is TypeScript","Elzero Web School","https://youtu.be/yUndnE-2osg","https://i.ytimg.com/vi/yUndnE-2osg/sddefault.jpg"
1,"Learn Typescript In Arabic 2022 - #02 - Install TypeScript And Transpile Files","Elzero Web School","https://youtu.be/pc5IlcEn8vw","https://i.ytimg.com/vi/pc5IlcEn8vw/sddefault.jpg"
2,"Learn Typescript In Arabic 2022 - #03 - Create Configuration And Watch Files","Elzero Web School","https://youtu.be/CSll1rsRPOI","https://i.ytimg.com/vi/CSll1rsRPOI/sddefault.jpg"
3,"Learn Typescript In Arabic 2022 - #04 - Statically vs Dynamically Typed Languages","Elzero Web School","https://youtu.be/OgxYA7G9HsM","https://i.ytimg.com/vi/OgxYA7G9HsM/sddefault.jpg"
4,"Learn Typescript In Arabic 2022 - #05 - Type Annotations And Any Data Type","Elzero Web School","https://youtu.be/quwf-YbyHVg","https://i.ytimg.com/vi/quwf-YbyHVg/sddefault.jpg"
5,"Learn Typescript In Arabic 2022 - #06 - Type Annotations With Arrays","Elzero Web School","https://youtu.be/U405xMeS4lM","https://i.ytimg.com/vi/U405xMeS4lM/sddefault.jpg"
6,"Learn Typescript In Arabic 2022 - #07 - Type Annotations With Multidimensional Arrays","Elzero Web School","https://youtu.be/n6JBmErg1OY","https://i.ytimg.com/vi/n6JBmErg1OY/sddefault.jpg"
7,"Learn Typescript In Arabic 2022 - #08 - Type Annotations With Function","Elzero Web School","https://youtu.be/ibvt_Ala8wE","https://i.ytimg.com/vi/ibvt_Ala8wE/sddefault.jpg"
8,"Learn Typescript In Arabic 2022 - #09 - Function Optional and Default Parameters","Elzero Web School","https://youtu.be/IS2VuO0IWso","https://i.ytimg.com/vi/IS2VuO0IWso/sddefault.jpg"

CHANGELOG.md

# CHANGELOG

## Todo

- add tests 📚

## Release

### 9 Oct 2022

- Reduce package size & files number
- `config` : find a better way to handel app version

### 8 Oct 2022

- Init Release 🎉


src/helpers/actions.js
import chalk from 'chalk';
import inquirer from 'inquirer';
import { oraPromise } from 'ora';
import config from './config.js';
import { exportInfoAsReport } from './export.js';
import {
	getPlaylistInfo,
	getPlaylistMetadata,
	getVideoInfo,
	getVideoMetadata,
} from './api.js';
import {
	getPlaylistId,
	getVideoID,
	handleApiError,
	validateApiKey,
	validateExportItems,
	validateFilePath,
} from './validate.js';

const playlistExportOptions = [
	{
		type: 'checkbox',
		name: 'exportItems',
		message: 'Which data do you want to export for each video?',
		choices: [
			{ name: '1. Position in the playlist', value: 'position' },
			{ name: '2. Title', value: 'title' },
			{ name: '3. Uploader', value: 'uploader' },
			{ name: '4. Uploader URL', value: 'uploaderUrl' },
			{ name: '5. URL', value: 'url' },
			{ name: '6. Description', value: 'description' },
			{ name: '7. Video privacy', value: 'videoPrivacy' },
			{ name: '8. Publish time (UTC)', value: 'publishTime' },
			{ name: '9. thumbnail', value: 'thumbnail' },
		],
		default: config.getExportOptionDefaults('playlistExportItems'),
		validate: validateExportItems,
	},
];

const videoExportOptions = [
	{
		type: 'checkbox',
		name: 'exportItems',
		message: 'Which data do you want to export for each video?',
		choices: [
			{ name: '1. Channel id', value: 'channelId' },
			{ name: '2. Channel title', value: 'channelTitle' },
			{ name: '3. Comment count', value: 'commentCount' },
			{ name: '4. Description', value: 'description' },
			{ name: '5. Favorite count', value: 'favoriteCount' },
			{ name: '6. Length', value: 'length' },
			{ name: '7. Likes', value: 'likeCount' },
			{ name: '8. Privacy status', value: 'privacyStatus' },
			{ name: '9. Publish date', value: 'publishDate' },
			{ name: '10. Tags', value: 'tags' },
			{ name: '11. Thumbnail', value: 'thumbnail' },
			{ name: '12. Title', value: 'title' },
			{ name: '13. Url', value: 'url' },
			{ name: '14. View count', value: 'viewCount' },
		],
		default: config.getExportOptionDefaults('videoExportItems'),
		validate: validateExportItems,
	},
];

const exportPathAndExtensionOptions = [
	{
		type: 'list',
		name: 'fileExt',
		message: 'Which file extension do you prefer?',
		choices: [
			{ name: 'JSON', value: 'json' },
			{ name: 'CSV', value: 'csv' },
		],
		default: config.get('fileExt'),
	},
	{
		type: 'input',
		name: 'folderPath',
		message: `An ${chalk.underline('absolute')} path of a ${chalk.underline(
			'folder'
		)} where the report will be saved:`,
		default: config.get('folderPath'),
		validate: validateFilePath,
	},
];

// ----------------
// api key action |
// ----------------

/**
 * ask user for apiKey and set it in config
 */
async function setApiKey() {
	const input = await inquirer.prompt([
		{
			type: 'input',
			name: 'apiKey',
			message: 'Enter your YouTube API key: ',
			validate: validateApiKey,
		},
	]);

	config.set('apiKey', input.apiKey.trim());
	console.log(chalk.green(`✔ Successfully set the API key`));
}

/**
 * Action handler for `tube-info key`
 */
async function apiKeyAction() {
	const apiKey = config.get('apiKey');

	if (apiKey) {
		console.log(`Your current YouTube API key is: ${chalk.yellow(apiKey)}`);

		const input = await inquirer.prompt([
			{
				type: 'list',
				name: 'keyAction',
				message: 'How would you like to manage your YouTube API key?',
				choices: [
					{ name: 'Edit key', value: 'editKey' },
					{ name: 'Remove key', value: 'removeKey' },
					{ name: 'Exit', value: 'exit' },
				],
			},
		]);

		switch (input.keyAction) {
			case 'editKey':
				setApiKey();
				break;
			case 'removeKey':
				config.set('apiKey', '');
				console.log(chalk.green(`✔ Successfully remove the API key`));
				break;
			default:
				break;
		}
	} else {
		console.log(chalk.red("✖ You haven't entered your YouTube API key!"));
		console.log(
			`  - Watch this 3 min. tutorial on how to get a YouTube API key (v3) for free - ${chalk.cyan(
				'https://youtu.be/N18czV5tj5o'
			)}`
		);
		setApiKey();
	}
}

// ---------------
// config action |
// ---------------

async function resetConfig() {
	const input = await inquirer.prompt([
		{
			type: 'confirm',
			name: 'resetConfig',
			message: 'Are you sure to reset all configurations to default?',
			default: false,
		},
	]);

	if (input.resetConfig) {
		config.resetAll();
		console.log(
			chalk.green('✔ Successfully reset all configurations to default.')
		);
	}
}

async function editPlaylistExportOptions() {
	const { exportItems } = await inquirer.prompt(playlistExportOptions);

	config.setExportOptionDefaults('playlistExportItems', exportItems);
}

async function editVideoExportOptions() {
	const { exportItems } = await inquirer.prompt([
		{
			type: 'checkbox',
			name: 'exportItems',
			message: 'Which data do you want to export for each video?',
			choices: [
				{ name: '1. Channel id', value: 'channelId' },
				{ name: '2. Channel Title', value: 'channelTitle' },
				{ name: '3. Comment Count', value: 'commentCount' },
				{ name: '4. Description', value: 'description' },
				{ name: '5. Favorite Count', value: 'favoriteCount' },
				{ name: '6. Length', value: 'length' },
				{ name: '7. Like Count', value: 'likeCount' },
				{ name: '8. Privacy Status', value: 'privacyStatus' },
				{ name: '9. Publish Date', value: 'publishDate' },
				{ name: '10. Tags', value: 'tags' },
				{ name: '11. Thumbnail', value: 'thumbnail' },
				{ name: '12. Title', value: 'title' },
				{ name: '13. Url', value: 'url' },
				{ name: '14. View Count', value: 'viewCount' },
			],
			default: config.getExportOptionDefaults('videoExportItems'),
			validate: validateExportItems,
		},
	]);

	config.setExportOptionDefaults('videoExportItems', exportItems);
}

async function editExportPathAndExtension() {
	const { fileExt, folderPath } = await inquirer.prompt(
		exportPathAndExtensionOptions
	);

	config.set('fileExt', fileExt);
	config.set('folderPath', folderPath);
}

async function editSkipPrivateOrDeleted() {
	const { skipPrivateOrDeleted } = await inquirer.prompt([
		{
			type: 'list',
			name: 'skipPrivateOrDeleted',
			message:
				'Do you want to skip/ignore any private or deleted videos in the playlist?' +
				"\nChoosing 'Skip' will NOT create an entry in the exported JSON/CSV file if the video is private/deleted.",
			choices: [
				{ name: 'Skip', value: true },
				{ name: 'Do not skip', value: false },
			],
			default: config.get('skipPrivateOrDeleted'),
		},
	]);

	config.set('skipPrivateOrDeleted', skipPrivateOrDeleted);
}

/**
 * Action handler for `tube-info config`
 * @param {{ path: boolean, reset: boolean }} options   Command options
 */
async function configAction(options) {
	if (options.path) {
		console.log(`Path of config file: ${chalk.cyan(config.path)}`);
		return;
	}

	if (options.reset) {
		resetConfig();
		return;
	}

	let exit = false;
	while (!exit) {
		const input = await inquirer.prompt([
			{
				type: 'list',
				name: 'configItem',
				message:
					'Which config do you want to edit? Choose "Exit" if you\'re done.',
				choices: [
					{
						name: 'Default playlist export options (eg. Items to export.)',
						value: 'playlistExportOptions',
					},
					{
						name: 'Default video export options (eg. title, thumbnail)',
						value: 'videoExportOptions',
					},
					{
						name: 'Export path & file extension',
						value: 'exportPathAndExtension',
					},
					{
						name: 'Skipping private or deleted videos',
						value: 'skipPrivateOrDeleted',
					},
					{ name: 'Exit', value: 'exit' },
				],
			},
		]);

		switch (input.configItem) {
			case 'playlistExportOptions':
				await editPlaylistExportOptions();
				console.log(chalk.green('✔ Saved playlist default export options.'));
				break;
			case 'videoExportOptions':
				await editVideoExportOptions();
				console.log(chalk.green('✔ Saved video default export options.'));
				break;
			case 'exportPathAndExtension':
				await editExportPathAndExtension();
				console.log(chalk.green('✔ Saved default export options.'));
				break;
			case 'skipPrivateOrDeleted':
				await editSkipPrivateOrDeleted();
				console.log(
					chalk.green('✔ Saved skipping private or deleted videos.\n')
				);
				break;
			case 'exit':
				exit = true;
				break;
		}
	}
}

// -----------------
// playlist action |
// -----------------

/**
 * Action handler for `tube-info id [options] <playlistId>`
 * @param {string} inputValue   Command argument - ID of playlist to be exported
 */
async function playlistAction(inputValue, options) {
	const playlistId = getPlaylistId(inputValue);

	// Check for "Watch Later"
	if (playlistId === 'WL') {
		console.log(
			chalk.yellow(
				'⚠️ Videos in Watch Later playlist cannot be retrieved through the YouTube API.'
			)
		);
		console.log(chalk.yellow('⚠️ Please try another playlist.'));
	}

	// Check if API key exists
	if (!config.get('apiKey')) {
		console.log(chalk.yellow(`⚠️ You haven't set your YouTube API key!`));
		console.log(`- Run ${chalk.cyan('tube-info key')} to set the API `);
		process.exit();
	}

	// Fetch playlist metadata
	const metadata = await oraPromise(
		getPlaylistMetadata(playlistId),
		'Fetching playlist metadata...'
	);

	console.log(`- Playlist Title: ${chalk.cyan(metadata.title)}`);
	console.log(
		`- Number of videos (including private videos): ${chalk.cyan(
			metadata.numOfVideos
		)}\n`
	);

	// Check if playlist is empty
	if (metadata.numOfVideos === 0) {
		console.log(
			chalk.yellow(
				'⚠️ This playlist is empty and there are no video data to export.'
			)
		);
	}

	let playlistData = null;

	const saveFileOptions = {
		fileExt: config.get('fileExt'),
		folderPath: config.get('folderPath'),
		title: metadata.title,
	};

	try {
		if (options.default) {
			// Skip all prompts for `--default` option
			const exportItems = config.getExportOptionDefaults('playlistExportItems');

			playlistData = await oraPromise(
				getPlaylistInfo(playlistId, exportItems),
				'Getting playlist info ...'
			);
		} else {
			const { exportItems } = await inquirer.prompt(playlistExportOptions);
			const { fileExt, folderPath } = await inquirer.prompt(
				exportPathAndExtensionOptions
			);

			saveFileOptions.fileExt = fileExt;
			saveFileOptions.folderPath = folderPath;

			// Fetch playlist data
			playlistData = await oraPromise(
				getPlaylistInfo(playlistId, exportItems),
				'Getting playlist info ...'
			);
		}
	} catch (error) {
		handleApiError(error);
	}

	exportInfoAsReport(playlistData, saveFileOptions);
}

// --------------
// video action |
// --------------

async function videoAction(inputValue, options) {
	const videoId = getVideoID(inputValue);

	// Check if API key exists
	if (!config.get('apiKey')) {
		console.log(chalk.yellow(`⚠️ You haven't set your YouTube API key!`));
		console.log(`- Run ${chalk.cyan('tube-info key')} to set the API `);
		process.exit();
	}

	const metaData = await oraPromise(
		getVideoMetadata(videoId),
		'Fetching video metadata...'
	);

	console.log(`- Video Title: ${chalk.cyan(metaData.title)}`);
	console.log(`- Channel title: ${chalk.cyan(metaData.channelTitle)}\n`);

	const saveFileOptions = {
		fileExt: config.get('fileExt'),
		folderPath: config.get('folderPath'),
		title: metaData.title,
	};

	let videoInfo = null;

	try {
		if (options.default) {
			// Skip all prompts for `--default` option
			const exportItems = config.getExportOptionDefaults('videoExportItems');

			videoInfo = await oraPromise(
				getVideoInfo(videoId, exportItems),
				'Getting video info ...'
			);
		} else {
			const { exportItems } = await inquirer.prompt(videoExportOptions);
			const { fileExt, folderPath } = await inquirer.prompt(
				exportPathAndExtensionOptions
			);

			saveFileOptions.fileExt = fileExt;
			saveFileOptions.folderPath = folderPath;

			// Fetch video data
			videoInfo = await oraPromise(
				getVideoInfo(videoId, exportItems),
				'Getting video info ...'
			);
		}
	} catch (error) {
		handleApiError(error);
	}

	exportInfoAsReport(videoInfo, saveFileOptions);
}

export { apiKeyAction, configAction, playlistAction, videoAction };
